### PR TITLE
Add callback to allow registration of feature renderers.

### PR DESCRIPTION
**Pull request by i509VCB**

### CODE REVIEW

File: fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/RegisterFeatureRendererCallback.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+@FunctionalInterface
+@Environment(EnvType.CLIENT)
+public interface RegisterFeatureRendererCallback {
+	Event<RegisterFeatureRendererCallback> EVENT = EventFactory.createArrayBacked(RegisterFeatureRendererCallback.class, callbacks -> (entityRenderer, acceptor) -> {
+		for (RegisterFeatureRendererCallback callback : callbacks) {
+			callback.registerFeatureRenderers(entityRenderer, acceptor);
+		}
+	});
+
+	void registerFeatureRenderers(LivingEntityRenderer<? extends LivingEntity, ? extends EntityModel<? extends LivingEntity>> entityRenderer, Consumer<FeatureRenderer<? extends LivingEntity, ? extends EntityModel<? extends LivingEntity>>> acceptor);
```
Comment (liach) [inline fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/RegisterFeatureRendererCallback.java]: can we create a local generic like `<T extends LivingEntity>` to simplify and guarantee some generics here

Comment (i509VCB) [reply fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/RegisterFeatureRendererCallback.java]: Compiler does not let you do that, it complains that `the method is generic`

Comment (liach) [reply fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/RegisterFeatureRendererCallback.java]: Really No for this?
```java
<T extends LivingEntity, M extends EntityModel<T>> void registerFeatureRenderers(LivingEntityRenderer<T, M> entityRenderer, Consumer<FeatureRenderer<T, M>> acceptor);
```

**Comment by JamiesWhiteShirt** [comment]
I worry about the usability of this event. Invoking it in the tail of the LivingEntityRenderer constructor means subclasses of LivingEntityRenderer will not be fully initialized when the event is invoked. Could this be an issue for consumers of this event?

**Comment by i509VCB** [comment]
@JamiesWhiteShirt I have changed how I approach the injection to where these events are fired after every feature renderer is registered.

@liach this is in render registries now (And I've worked around most of the generics issues)

@sfPlayer1 A few internal changes, would like you to look at this again

File: fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/RegisterFeatureRendererCallback.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	/**
+	 * Called when feature renderers may be registered.
+	 *
+	 * @param entityType     the entity type of the renderer
+	 * @param entityRenderer the entity renderer
+	 * @param acceptor       the accepted used to register feature renderers
+	 */
+	void registerFeatureRenderers(EntityType<? extends LivingEntity> entityType, LivingEntityRenderer<?, ?> entityRenderer, FeatureAcceptor acceptor);
+
+	final class FeatureAcceptor {
```
Comment (liach) [inline fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/RegisterFeatureRendererCallback.java]: imo you should just make this an interface and the impl creates an instance of `FeatureAcceptor` when it calls the event instead.

Comment (i509VCB) [reply fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/RegisterFeatureRendererCallback.java]: I'll consider it. Could even implement it onto living entity renderer itself if I went for an interface (but that would not be part of docs and an impl detail

Comment (liach) [reply fabric-renderer-registries-v1/src/main/java/net/fabricmc/fabric/api/client/rendereregistry/v1/RegisterFeatureRendererCallback.java]: Exposing this in the api defeats the point of abstraction; users don't need to know how this interface is implemented.

File: fabric-renderer-registries-v1/src/testmod/java/net/fabricmc/fabric/test/renderer/registry/FeatureRendererTests.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+import net.minecraft.client.render.entity.model.PlayerEntityModel;
+import net.minecraft.client.util.math.MatrixStack;
+import net.minecraft.entity.EntityType;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.decoration.ArmorStandEntity;
+
+import net.fabricmc.api.ClientModInitializer;
+import net.fabricmc.fabric.api.client.rendereregistry.v1.RegisterFeatureRendererCallback;
+
+public class FeatureRendererTests implements ClientModInitializer {
```
Comment (modmuss50) [inline fabric-renderer-registries-v1/src/testmod/java/net/fabricmc/fabric/test/renderer/registry/FeatureRendererTests.java]: Unless im blind I dont see this being added to the test mod json?

**Review by modmuss50 — CHANGES_REQUESTED** [review]
Just the test mod that confuses me?

File: fabric-renderer-registries-v1/src/testmod/java/net/fabricmc/fabric/test/renderer/registry/FeatureRendererTests.java (hunk 3 above)
Comment (i509VCB) [reply fabric-renderer-registries-v1/src/testmod/java/net/fabricmc/fabric/test/renderer/registry/FeatureRendererTests.java]: This testmod was primarily designed for testing generics on the methods. I'll hook up an actual simple feature to render in a bit.

**Review by liach — APPROVED** [review]

**Review by Player3324 — APPROVED** [review]
LGTM, thanks

For other reviewers:
The class name choice is difficult, but considering the niche aspect and ambiguity (worldgen uses feature too), I believe the lengthy name is justified.
We were also unable to improve upon the highly wildcarded registerRenderers  signature. Tighter type constraints only added noise and caused trouble with lambdas or other uses.

**Review by modmuss50 — APPROVED** [review]
Look good, can be moved to last call.

I think the class name is fine as like said it could easily clash with other things.

Will move to last call.

**Review by hYdos — APPROVED** [review]
looks good

**Pull request merged**